internal/provider/resource_access_offer.go

```diff
@@ -0,0 +1,436 @@
+				Access:   "read",
+				OfferURL: offerURL,
+			})
+			if err != nil {
+				resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to update access offer resource, got error: %s", err))
+				return
+			}
+		}
+	}
+	// Check if is a new user and grant access
+	// If not, revoke all access and grant the right one
+	for _, planUser := range planUsers {
+		if _, ok := offerUserNames[planUser]; !ok {
+			// user has no access yet so we need to grant
+			a.trace(fmt.Sprintf("grant permission %q to user %q for offer %q", access, planUser, offerURL))
+			err := a.client.Offers.GrantOffer(&juju.GrantRevokeOfferInput{
+				User:     planUser,
+				Access:   access,
+				OfferURL: offerURL,
+			})
+			if err != nil {
+				resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to create access offer resource, got error: %s", err))
+				return
+			}
+			continue
+		}
+		// revoke read (remove all)
+		a.trace(fmt.Sprintf("revoke user %q for offer %q", planUser, offerURL))
+		err := a.client.Offers.RevokeOffer(&juju.GrantRevokeOfferInput{
+			User:     planUser,
+			Access:   "read",
+			OfferURL: offerURL,
+		})
+		if err != nil {
+			resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to update access offer resource, got error: %s", err))
+			return
+		}
+		// grant the right permission
+		a.trace(fmt.Sprintf("grant permission %q to user %q for offer %q", access, planUser, offerURL))
+		err = a.client.Offers.GrantOffer(&juju.GrantRevokeOfferInput{
+			User:     planUser,
+			Access:   access,
+			OfferURL: offerURL,
+		})
+		if err != nil {
+			resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to update access offer resource, got error: %s", err))
+			return
+		}
+	}
+	plan.ID = types.StringValue(newAccessOfferIDFrom(offerURL, access, planUsers))
+	plan.Access = types.StringValue(access)
+	newStateUsers, errDiag := basetypes.NewSetValueFrom(ctx, types.StringType, planUsers)
+	plan.Users = newStateUsers
+	resp.Diagnostics.Append(errDiag...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	resp.Diagnostics.Append(resp.State.Set(ctx, &plan)...)
+}
+
+func (a *accessOfferResource) Delete(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse) {
+	// Check first if the client is configured
+	if a.client == nil {
+		addClientNotConfiguredError(&resp.Diagnostics, "access offer", "read")
+		return
+	}
+	var plan accessOfferResourceOffer
+
+	// Get the Terraform state from the request into the plan
+	resp.Diagnostics.Append(req.State.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Get the users
+	var users []string
+	resp.Diagnostics.Append(plan.Users.ElementsAs(ctx, &users, false)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Revoking against "read" guarantees that the entire access will be removed
+	// instead of only decreasing the access level.
+	for _, user := range users {
+		err := a.client.Offers.RevokeOffer(&juju.GrantRevokeOfferInput{
```

Comment:
this might also not be completely valid thing to do - really depends on other resource_access_offer resources in the plan.